tools/database/src/analyse_gas_usage.rs

```diff
@@ -0,0 +1,348 @@
+use nearcore::open_storage;
+
+use crate::block_iterators::{CommandArgs, LastNBlocksIterator};
+
+/// `Gas` is an u64, but it stil might overflow when analysing a large amount of blocks.
+/// 1ms of compute is about 1TGas = 10^12 gas. One epoch takes 43200 seconds (43200000ms).
+/// This means that the amount of gas consumed during a single epoch can reach 43200000 * 10^12 = 4.32 * 10^19
+/// 10^19 doesn't fit in u64, so we need to use u128
+/// To avoid overflows, let's use `BigGas` for storing gas amounts in the code.
+type BigGas = u128;
+
+#[derive(Parser)]
+pub(crate) struct AnalyseGasUsageCommand {
+    /// Analyse the last N blocks in the blockchain
+    #[arg(long)]
+    last_blocks: Option<u64>,
+
+    /// Analyse blocks from the given block height, inclusive
+    #[arg(long)]
+    from_block_height: Option<BlockHeight>,
+
+    /// Analyse blocks up to the given block height, inclusive
+    #[arg(long)]
+    to_block_height: Option<BlockHeight>,
+}
+
+impl AnalyseGasUsageCommand {
+    pub(crate) fn run(&self, home: &PathBuf) -> anyhow::Result<()> {
+        // Create a ChainStore and EpochManager that will be used to read blockchain data.
+        let mut near_config =
+            nearcore::config::load_config(home, near_chain_configs::GenesisValidationMode::Full)
+                .unwrap();
+        let node_storage: NodeStorage = open_storage(&home, &mut near_config).unwrap();
+        let store: Store =
+            node_storage.get_split_store().unwrap_or_else(|| node_storage.get_hot_store());
+        let chain_store = Rc::new(ChainStore::new(
+            store.clone(),
+            near_config.genesis.config.genesis_height,
+            false,
+        ));
+        let epoch_manager =
+            EpochManager::new_from_genesis_config(store, &near_config.genesis.config).unwrap();
+
+        // Create an iterator over the blocks that should be analysed
+        let blocks_iter_opt = crate::block_iterators::make_block_iterator_from_command_args(
+            CommandArgs {
+                last_blocks: self.last_blocks,
+                from_block_height: self.from_block_height,
+                to_block_height: self.to_block_height,
+            },
+            chain_store.clone(),
+        );
+
+        let blocks_iter = match blocks_iter_opt {
+            Some(iter) => iter,
+            None => {
+                println!("No arguments, defaulting to last 100 blocks");
+                Box::new(LastNBlocksIterator::new(100, chain_store.clone()))
+            }
+        };
+
+        // Analyse
+        analyse_gas_usage(blocks_iter, &chain_store, &epoch_manager);
+
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+struct GasUsageInShard {
+    pub used_gas_per_account: BTreeMap<AccountId, BigGas>,
+    pub used_gas_total: BigGas,
+}
+
+/// A shard can be split into two halves.
+/// This struct represents the result of splitting a shard at `split_account`.
+#[derive(Debug, Clone)]
+struct ShardSplit {
```

Comment:
Now looking at the updated results, is the split account significant? If so can you print the data in the same format as the tooling for analysing storage? The format is roughly left:middle:right where each is a percentage value e.g. 40%:20%:40%.

Reply:
>  is the split account significant?

I thought that it might be. `split_account` is the first account that made the left half too big and started the right half, so it could be some sort of big account that tipped the scale over.

Looking at the actual results, this is not the case. All `split_accounts` have low gas usage. Because of that I don't know if there is a point in displaying `left:middle:right`, as the middle will just be `0%`.

Reply:
Nevermind, in the new version (where left is defined as <split instead of <=split) the split accounts are actually big. Add the `left:middle:right` display.